From c0620c68229bace31d506f3b04d3385cb9761969 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Sat, 21 Sep 2024 12:08:44 +0530
Subject: [PATCH] wifi: ath12k: Fix ppdu memset issue to avoid stale values

Currently, the mpdu_retry and errmap members are located after the
cmn_mpdu_info member in the struct hal_rx_mon_ppdu_info. When clearing the
existing ppdu_info to avoid a bulk length memset, the size is calculated
by subtracting the size of the ppdu_info struct from the size of the
cmn_mpdu_info struct. Because mpdu_retry and errmap are at the end, they
contain junk values, and the first 64 bits of cmn_mpdu_info are
unintentionally cleared. To fix this, move mpdu_retry and errmap
before the cmn_mpdu_info member.

Fixes : (I1c4bf9c0f36) wifi: cfg80211/ath12k: add support to count rx retries
Fixes : (Ie4f1db52c9) wifi: mac80211/ath12k: add support to ethtool stats

Patch-dependency: 869-b-wifi-ath12k-add-support-to-count-rx-retries.patch

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/hal_rx.h | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/hal_rx.h b/drivers/net/wireless/ath/ath12k/hal_rx.h
index f91e78e..fb1e80e 100644
--- a/drivers/net/wireless/ath/ath12k/hal_rx.h
+++ b/drivers/net/wireless/ath/ath12k/hal_rx.h
@@ -290,10 +290,12 @@ struct hal_rx_mon_ppdu_info {
 	u32 eht_known;
 	u32 eht_data[9];
 	u32 eht_user_info[EHT_MAX_USER_INFO];
-	struct hal_rx_u_sig_info u_sig_info;
-	struct hal_rx_mon_cmn_mpdu_info cmn_mpdu_info;
 	bool mpdu_retry;
 	u32 errmap;
+	struct hal_rx_u_sig_info u_sig_info;
+
+	/* must be last */
+	struct hal_rx_mon_cmn_mpdu_info cmn_mpdu_info;
 };
 
 #define HAL_RX_PPDU_START_INFO0_PPDU_ID		GENMASK(15, 0)
-- 
2.34.1

